1**. Найти с помощью NumPy SVD для матрицы

$$\begin{pmatrix}
1 & 2 & 0\\ 
0 & 0 & 5\\ 
3 & -4 & 2\\ 
1 & 6 & 5\\ 
0 & 1 & 0
\end{pmatrix}.$$




In [4]:
import numpy as np

In [15]:
A = np.array([[1, 2, 0],
              [0, 0, 5],
              [3, -4, 2],
              [1, 6, 5],
              [0, 1, 0]])
print(f'Матрица A:\n{A}')

Матрица A:
[[ 1  2  0]
 [ 0  0  5]
 [ 3 -4  2]
 [ 1  6  5]
 [ 0  1  0]]


In [6]:
U, s, W = np.linalg.svd(A)

# Транспонируем матрицу W
V = W.T

# s - список диагональных элементов, его нужно привести к виду диагональной матрицы для наглядности
D = np.zeros_like(A, dtype=float)
D[np.diag_indices(min(A.shape))] = s

In [7]:
print(f'Матрица D:\n{D}')

Матрица D:
[[8.82486885 0.         0.        ]
 [0.         6.14060608 0.        ]
 [0.         0.         2.53271528]
 [0.         0.         0.        ]
 [0.         0.         0.        ]]


In [8]:
print(f'Матрица U:\n{U}')

Матрица U:
[[ 0.17056501  0.15680918 -0.53077508 -0.79905375 -0.16158397]
 [ 0.39287016 -0.52933945  0.6134793  -0.43375771  0.03082495]
 [-0.14366152 -0.82449256 -0.52379105  0.14049848  0.07400343]
 [ 0.88843702  0.06074346 -0.24655277  0.37755832 -0.06042632]
 [ 0.08125046  0.10831843 -0.08231425 -0.10524851  0.98173958]]


In [9]:
# Убедимся, что она действительно ортогональна
print(np.dot(U.T, U))

[[ 1.00000000e+00  3.27862737e-16 -3.91180144e-16  2.08166817e-17
  -2.08166817e-17]
 [ 3.27862737e-16  1.00000000e+00  2.96637714e-16  9.88792381e-17
   3.12250226e-17]
 [-3.91180144e-16  2.96637714e-16  1.00000000e+00 -1.61329283e-16
  -4.85722573e-17]
 [ 2.08166817e-17  9.88792381e-17 -1.61329283e-16  1.00000000e+00
  -2.77555756e-17]
 [-2.08166817e-17  3.12250226e-17 -4.85722573e-17 -2.77555756e-17
   1.00000000e+00]]


In [10]:
print(f'Матрица V:\n{V}')

Матрица V:
[[ 0.07116451 -0.36737824 -0.92734505]
 [ 0.71702467  0.66514082 -0.20847855]
 [ 0.69340553 -0.65009301  0.31075368]]


In [11]:
# Убедимся, что она действительно ортогональна
print(np.dot(V.T, V))

[[ 1.00000000e+00  2.77555756e-16 -5.55111512e-17]
 [ 2.77555756e-16  1.00000000e+00 -1.38777878e-16]
 [-5.55111512e-17 -1.38777878e-16  1.00000000e+00]]


In [12]:
# Проведем проверку
print(np.dot(np.dot(U, D), V.T))

[[ 1.00000000e+00  2.00000000e+00  1.33226763e-15]
 [ 2.22044605e-16 -2.60902411e-15  5.00000000e+00]
 [ 3.00000000e+00 -4.00000000e+00  2.00000000e+00]
 [ 1.00000000e+00  6.00000000e+00  5.00000000e+00]
 [-2.77555756e-17  1.00000000e+00 -4.16333634e-17]]


2. Для матрицы из предыдущего задания найти:

    а) евклидову норму;
    
    б) норму Фробениуса.

In [13]:
from numpy.linalg import norm

In [16]:
print(f'l2 Евклидова норма вектора А:\n{norm(A, ord=2)}')

l2 Евклидова норма вектора А:
8.824868854820444


In [26]:
print(f'Первый диагональный элемент матрицы D: {D[0][0]}')

Первый диагональный элемент матрицы D: 8.824868854820444


In [21]:
print(f'Фробениуса норма вектора А:\n{norm(A, ord="fro")}')

Фробениуса норма вектора А:
11.045361017187261


Определим Фробениуса норму вектора $a$:

$$\left\|x\right\|_{2} = \sqrt{1^{2}+2^{2}+(0)^{2}+0^{2}+0^{2}+(5)^{2}+3^{2}+(-4)^{2}+(2)^{2}+1^{2}+6^{2}+(5)^{2}+0^{2}+1^{2}+(0)^{2}}=\sqrt{122}=11.045.$$